pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -176,6 +176,11 @@ func calcRestartCountByLogDir(path string) (int, error) {
 // * start the container
 // * run the post start lifecycle hooks (if applicable)
 func (m *kubeGenericRuntimeManager) startContainer(ctx context.Context, podSandboxID string, podSandboxConfig *runtimeapi.PodSandboxConfig, spec *startSpec, pod *v1.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, podIP string, podIPs []string) (string, error) {
+	// Don't let the image garbage collection race with container creation
+	// See: https://github.com/kubernetes/kubernetes/issues/123631
+	m.imageGCLock.RLock()
+	defer m.imageGCLock.RUnlock()
+
```

Comment:
I think we do need this, but I wonder if the lock scope is correct. I worry this function will take a while and it will delay GC a dangerous amount

Reply:
That's a good point. I moved the lock below `imagePuller.EnsureImageExists` which leaves a small gap of code to be racing within that method, but I assume the pull itself is the biggest part in the container creation process.